### DISCUSSION TITLE
Ideas about how to handle Socket REPL in the future

### DISCUSSION THREAD

**Discussion by mauricioszabo** [mauricioszabo/atom-chlorine#241] [General]
Right now, Chlorine adds a blob of Clojure code to Socket REPL and then "transforms it" into a programmable REPL. It uses UNREPL. This gives us good support for tooling, but there are some inputs that can break the whole REPL (notably, Datahike library) and also it's quite a pain to support the more advanced features that I want to add like interactive renders, etc.

So, I'm thinking on some alternatives for newer versions. I'm thinking about adding two new configs: one for REPL Mode and one from Printer Mode (I need to rethink these names to be more self-explanatory), but the idea is that _REPL Mode_ is how the REPL behave, and _Printer Mode_ is how the results will be wrapped/rendered on Chlorine. 

So, for available REPL Modes, I'm thinking on (ordered for the "simpler one" to "more advanced"):

## Fully unstructured REPL ("Inferior REPL mode")

This would just evaluate forms, and the very next code that returns will be understood as the output. So, if someone writes `(prn {:a 10 :b 20})`, Chlorine will think that the map is the "result" and that `nil` (the result of `prn`) is the "stdout". This is obviously very crude, but it can open the door for great things like connecting to more socket-repl-like elements that would capture the output (I believe Reveal have an option to work as a socket REPL too), or toolings like https://github.com/seancorfield/socket-rebl

## Rewrite code before sending to REPL ("Wrapped input mode")

This will wrap the code in something like `[tooling$repl-result <some-id> "string-with-the-result"]` and then try to use this EDN format to detect what is an output and what is not. This is already how things are done in environments like Clojerl, Joker, etc.

## Upgrade the REPL to be programmable ("Programmable REPL mode")

This will do exactly the same thing we're doing today with UNREPL, but with a simple alternative that maybe is more portable. This will be the _only way_ to have "break evaluation" on some environments, and I'm thinking if it's possible to support Clojure, Clojure.CLR, and maybe Clojerl, Lumo, and Plank. Unfortunately, ClojureScript REPLs will never support break evaluation...

## Specific REPL

This is not a different mode _per se_, but I'm thinking about adding full-class support for `prepl` (so maybe Figwheel REPL can be used) and Shadow remote API (for both Clojure *and* ClojureScript). I'm not yet sure about this later option, so maybe I'll add a specific config for Shadow (like "always use shadow-remote") because I believe there are people that may want to use a different "REPL Mode" for Clojure and ClojureScript.

---

So, if someone says that he wants to use "Programmable REPL" and that option is not available on his REPL, he would fallback to the "Wrapped input" for example.

**Comment by sogaiu** [comment]
Thank you for this write-up!

**Reply by seancorfield** [reply]
Full first-class support for `prepl` would allow for Chlorine/Clover to be used as a client to the remote-repl setup that Reveal offers, which means you can use Reveal as a proxy between the client (editor) and the server on which code runs. That lets you use Reveal in front of a ClojureScript REPL for example, or a remote JVM process, or whatever. See https://vlaaad.github.io/reveal-repls-and-networking
